lib/pbench/test/unit/server/query_apis/commons.py

```diff
@@ -159,38 +162,34 @@ def missing_key_helper(keys):
         for r in range(1, len(parameter_items) + 1):
```

Comment:
You should probably start the range a zero, to test the empty case:
```
        for r in range(len(parameter_items) + 1):
```

Reply:
Empty payload test is different and it behaves differently, the response would be `Invalid request payload` which already has its own test function `test_missing_json_object`, for the missing keys test we need at least one key in the payload and the response would be `Missing required parameters: <parameters>`

Reply:
I think it's different:  I think what I'm describing will send a payload which is an object which contains no fields; I think what you are describing either sends no payload or sends a payload which contains no object (I'm not sure exactly what the difference between those might be).  So, I think what I'm describing is still a valid test which is different from what you're already testing.

Reply:
you mean sending json object as `{}`? 
```
response = client.post(
                server_config.rest_uri + self.pbench_endpoint,
                headers={"Authorization": "Bearer " + pbench_token},
                json={},
            )
```
If this is what you mean then its as good as sending no payload

Reply:
"As good as" perhaps, but nevertheless _different_.  That is, a payload of `{}` has a length of `2` and a type of `application/json`, whereas "no payload" has a length of zero, and I suspect that the type is never used.  I would have to dive deep into the `requests` or Flask documentation to explore the difference, but it would probably be pretty straightforward for you to just crank up the server and try it.

It's possible that our code doesn't discriminate between the two cases.  And, it's possible that the layers above/below our code paper over the difference for us.  My knowledge isn't that detailed.  But, given that we hope to support multiple clients, it would be good to ensure that we handle both cases gracefully.

Reply:
Yes I experimented with it and we validate the json data in our post and get request (schema validation) before proceeding where it treats `{}` and `None` as same.

If you look in the class `Schema` and `validate` function line 363 in `lib/pbench/server/api/resources/query_apis/__init__.py` you will see we raise `InvalidRequestPayload()` if no json data present
```
if not json_data:
    raise InvalidRequestPayload()
```

Reply:
Yes, we do fail out with an empty payload. I'm not sure I can think of a "query api" that would have no JSON input, but then you've already got the index mappings where the parameter is in the URI rather than as JSON. If you'd based it on `ElasticBase`, you'd be tripping over this code.

Arguably, the `Schema` validator should fail with an empty or missing payload only if the schema has one or more required parameters. It's too bad we can't iterate over `None`, because we'd be fine if we just let it complain about missing required parameters later.

Reply:
I've lost my grip, and I think the time for this discussion has passed, so I'm just going to resolve it.